### ISSUE TITLE
Long Term Statistics

### ISSUE THREAD

**Issue by jjhuff** [neggert/hass-egauge#44]
Thanks for the integration!

With 2021.8 out, it'd be awesome to be able to integrate my eGauge system into HAs new energy management framework. They don't show up in the list of available options currently.

It looks like there might be something missing with respect to https://developers.home-assistant.io/docs/core/entity/sensor/#long-term-statistics.

**Comment by steppel** [comment]
With update to 2021.8 mine doesn't work anymore. I deinstalled and now I can't install it anymore. In Configuration/Integrations/Add Integration/eGauge .. I get error Config flow could not be loaded.

**Comment by jdwhite** [comment]
I had to re-install using HACS and select version 0.2.0. Stock add-ons seems to be serving the 0.1.0 equivalent.

**Comment by neggert** [comment]
I'm aiming to have the changes to support long-term statistics done sometime in the next week. Looks like it should be straightforward.

There's another issue going on where HACS thinks v0.2.0 is older than v0.1.0. Not sure what's going on there, but I'm investigating.

**Comment by neggert** [comment]
Released v0.2.1 with minimal changes and HACS seems to be picking up the correct version now 🤷 .

**Comment by steppel** [comment]
Awesome, could install and it's workings. Thanks for the quick turnaround. Looking forward to the next update so I can add the eGauge to the energy tab in HA! 

Thanks for all the work, appreciate it!

**Comment by neggert** [comment]
The existing "rolling" counters won't really work with the new energy dashboard. The new Hass functionality is designed for a counter that resets on some regular interval, e.g. at midnight every day. The current counters show usage in e.g. the last 24 hours, rather than usage since midnight.

So, long story short, I made some changes in master that add a new sensor reporting "today's" usage (i.e. energy used since midnight local time). The new sensor has all the attributes needed to use with the new energy dashboard.

This seems to be working on my system, but I could use another person or two to test it out. You can update to the code that's on the master branch by selecting "re-install" in HACS, then choosing "master" for the version.

**Comment by steppel** [comment]
<img width="1796" alt="Screen Shot 2021-08-06 at 4 40 46 PM" src="https://user-images.githubusercontent.com/49464956/128573904-43d83fad-0e0c-4ad1-b782-a85e6bf0397d.png">

Yep shows up! I will keep an eye on it over the coming days and compare with the eGauge page.

**Comment by steppel** [comment]
Something is off. 

![image](https://user-images.githubusercontent.com/49464956/128605008-3b223423-3eed-42d4-9b73-e027d9631c36.jpeg)

![image](https://user-images.githubusercontent.com/49464956/128605011-c1a745b2-d9bb-40b9-b448-96bcc80c6b2d.png)

I produce more than I use at the moment. But the energy tab shows I'm consumed (movement of the dots from grid to home). Not sure what going on but I would expect to see a positive number on the home icon for consumption, generation and grid and just the dots flowing different directions. It doesn't seem to match what eGauge page is displaying. But maybe I set it up wrong? Tried all 5 sensors.

**Comment by jdwhite** [comment]
I have to keep reminding myself that the energy page isn't showing instantaneous _power_ usage, but _energy_ usage over the span of a day.  I find it curious that your "today's use" number is negative -- mine is positive.  I realize you can have a net negative usage (gross - generated) but seems like "today's use" should be a gross increasing counter of today's consumption in kWh.

I'm thinking the integration needs a "return to grid" sensor -- for the return to grid configuration in the energy configuration -- and would be the difference between today's 'use' and 'gen', so may be possible to create a template sensor for that. I'm not sure if it would have the proper attributes to make HASS show it as a valid entity for selection.  I would expect to see a line showing that I've pushed ~1 kWh back to the grid, but I don't.

Obligatory screenshot. Note my pre-9am graph columns may be goofy due to be configuring "todays gen" as "return to grid" overnight.  Not sure why I have a "negative" energy of 1 kWh usage overnight. Possibly just need to wait until tomorrow and see if things shake out.

![hass-energy 2021-08-07_11](https://user-images.githubusercontent.com/1943283/128606350-a8154008-2a11-4c4f-ac5a-7f53470cb775.png)

**Comment by steppel** [comment]
Can you make sensor.egauge_grid and egauge_solar available? The today values don't work for the energy tab (yes they show up but it's not reflecting correct)

**Comment by steppel** [comment]
Wouldn't it make more sense to use egauge_grid and egauge_solar ![image](https://user-images.githubusercontent.com/49464956/128607982-01de2be0-8c5e-416d-aabc-3510c1eec33d.png)

**Comment by jdwhite** [comment]
I don't believe so as those are instantaneous sensors. As I understand it, Energy expects cumulative counters that reset at midnight. At 12 minutes past each hour (according to Paulus on last Wednesday's release party livestream), statistics are regenerated and stored.

**Comment by phidauex** [comment]
To have an instantaneous sensor work with the statistics module, it looks like you need to set `state_class` to `measurement`.

https://developers.home-assistant.io/docs/core/entity/sensor/#long-term-statistics

That seems to imply to me that the instantaneous grid and solar sensors would work, we just need to let HASS know that it is an instantaneous measurement, and then it will track the statistics. This would somewhat duplicate the metered entities, but it seems like it could work side by side.

**Comment by neggert** [comment]
It took a little bit of experimentation to figure this out, but instantaneous (power) sensors do *not* directly work with the new energy dashboard. To show up in the dashboard, the sensors must meet three requirements:

1 `state_class` set to "measurement"
2 `last_reset` returns *something*
3 `unit_of_measurement` in kWh

In the current master branch code, instantaneous registers will have (1), but not (2) or (3), so they can't be directly added to the energy dashboard.

I think if you wanted to use the instantaneous values, you'd need to feed them into the [integration](https://www.home-assistant.io/integrations/integration/) module to convert from power to energy (the Home Assistant blog post doesn't mention this step, which I think is confusing a lot of people, myself included). There's no reason to do that with an eGauge, because the eGauge historical registers already do the integration for you, and sample at a much higher rate than you'd get by doing this in hass.

So that's how to get value into the dashboard, but it's also a little bit tricky to get the dashboard to display the correct values, because the dashboard is expecting a different set of values than most eGauges are set up to measure. The dashboard wants 3 values:

* Energy drawn from grid
* Energy produced to grid
* Energy generated

All three of these are expected to be positive numbers.

My eGauge is set up to measure two values. Not sure if this is standard, since you can really set up an eGauge however you want, but I assume so. My system measures:

* Energy generated (`gen`)
* Total energy used (solar + grid) (`use`)

So to get this working, we need to do some calculations. I did this by creating new registers in the eGauge, but it might be possible to do this with templates.

* Energy drawn from grid: `grid_in = max(0, use - gen)`
* Energy produced to grid: `grid_out = max(0, gen - use)`

Here's how my dashboard looks after doing this:

![image](https://user-images.githubusercontent.com/1271014/128642450-d9737f27-13d4-4b4c-b3ef-649e5bd0cb65.png)

Setting this up on the eGauge itself, was a little hairy, so I'm going to investigate some options for the easiest way to set this up.

**Comment by neggert** [comment]
A few more data points:

* [Templates are not supported (yet)](https://community.home-assistant.io/t/energy-management-in-home-assistant/326854/30?u=cannedbass)
* Seem that my install is probably *not* standard

Here's the standard install from the eGauge manual. I marked the location of my CT1 and CT2 in read. How is everyone else set up?

![image](https://user-images.githubusercontent.com/1271014/128643001-c22928dc-a95e-4d93-ab61-ae5f2f230a26.png)

**Comment by neggert** [comment]
It looks like setting this up on the eGauge itself is the best option. There are a couple of ways to do this depending on your setup. We'll probably need to add instructions for this to the README.

This is my current understanding of how to set this up. When in doubt, refer to the [config reference](https://www.egauge.net/media/support/docs/config-guide.pdf). We're going to make changes in the menu under Settings->Install on the web interface.

There are two common setups:

## Standard Split-Phase back-fed

![image](https://user-images.githubusercontent.com/1271014/128644739-dd24b5a1-c942-42ac-8607-ce121b4fce6c.png)

Here, you have a sensor measuring power to and from the grid. All we need to do is to make sure that the eGauge is recording versions of this measurement that record energy in and out, in addition to the net. To do this, we just need to tick a few checkboxes:

![image](https://user-images.githubusercontent.com/1271014/128644813-0a67c836-9c8b-4fa5-85fe-970ad02b0128.png)

This should create a couple new sensors in Hass: "egauge todays grid+" and "egauge todays grid-". You can use these for "Grid consumption" and "Returned to Grid" in the energy dashboard.

Note: I don't have a system set up this way, so this is just my best guess. I'd love someone to confirm that this works.

## Direct-Feed Solar

![image](https://user-images.githubusercontent.com/1271014/128644929-009373c2-cc1f-4862-9e3f-0357f130f1f6.png)

In this setup, we have sensors measuring solar production and total usage by the main panel, but we have to do a little math to figure out how much we're drawing from or producing to the grid. We can do this using template sensors in eGauge. Note I think you need a recent version of the eGauge firmware for this to work. I'm on 4.2.5.

![image](https://user-images.githubusercontent.com/1271014/128645023-d50a04b9-4c02-4681-ab5b-3d8880a09bfa.png)

This should create a couple new sensors in Hass: "egauge todays grid_n" and "egauge todays grid_out". You can use these for "Grid consumption" and "Returned to Grid" in the energy dashboard.

## Other

Refer to the [config guide](https://www.egauge.net/media/support/docs/config-guide.pdf) and try to figure out how to create registers that give you you need. Please post here if you get it working so that others can learn.

**Comment by steppel** [comment]
![image](https://user-images.githubusercontent.com/49464956/128647319-478a5eda-de6f-4be0-9035-cff948749444.png)


When I click the negative value and try to save I get that message.

**Comment by steppel** [comment]
I get an error when trying to add the negative value.

![image](https://user-images.githubusercontent.com/49464956/128647348-5bc411a9-64b7-48f4-8509-6d999e997579.jpeg)

**Comment by steppel** [comment]
![image](https://user-images.githubusercontent.com/49464956/128647494-6e30be36-aac1-4f21-a879-cd32fc94ea6c.jpeg)


That's what it is right now and HA shows.

![image](https://user-images.githubusercontent.com/49464956/128647508-386d21bd-5019-4b0c-a51f-9cc55abc2f07.jpeg)


Seems off, is it supposed to show the as eGauge ?

**Comment by neggert** [comment]
Hmm, that's annoying that it won't let you select "=+-". I'll have to think about how to work around that one. Suggestions welcome.

I think the energy distribution graph shows totals for the day. So it should match what you see in eGauge if you select "Summary since start of day" in the green box above the graph. BTW, I've noticed strange numbers in the hass dashboard when I've been fiddling with things. To check that everything is working reliably, I'd suggest letting it run for a full day without touching anything.

**Comment by jdwhite** [comment]
@neggert - Thanks for your diligent work on this. I have a direct feed solar setup, and used your instructions to create two new registers. I have an EG3000 with firmware v4.2.5 and had no problems with your instructions. After reinstalling master from HACS I can see the new sensors and have configured them accordingly in energy. Right now my solar production is a negative value -- always has been despite the sensor I'm using having a positive value, so not sure what the deal is there but we'll see what tomorrow brings. Right now the to/from grid sensors are nearly zero since they were just created so nothing much to see.

**Comment by steppel** [comment]
Mine stopped showing anything since today. Didn't change anything just had one reboot of HA yesterday. All sensors unavailable but eGauge is online and running.

![image](https://user-images.githubusercontent.com/49464956/128760056-f3ffd5b0-e5fb-4539-ad30-f6bc6fda1ab3.png)

**Comment by neggert** [comment]
I think this might be the best way to do the setup for a back-fed system. Let me know if it works.

![image](https://user-images.githubusercontent.com/1271014/128790115-25267474-414a-4c63-b63e-3d8b1a5b1737.png)

**Comment by jdwhite** [comment]
@neggert - I just saw your last response as I was putting together a response of what I observed today. I will wait on most of that post but one thing I did encounter is that a number of the sensors seem to be reporting Wh instead of Kwh. 

My current eGauge config:
![egauge_registers](https://user-images.githubusercontent.com/1943283/128793657-b0421270-4e1a-4264-bce7-4d3da7b07997.png)

What I see in Home Assistant:

![image](https://user-images.githubusercontent.com/1943283/128793736-4ca29d8a-6222-4976-87b7-9fbc5fbe034c.png)
![image](https://user-images.githubusercontent.com/1943283/128793787-288d86f3-3015-4f03-9243-6272bc31291f.png)
![image](https://user-images.githubusercontent.com/1943283/128793883-d35b17da-0548-47f7-99dd-bf0bf6f43be3.png)

The rest of my post was going to be about how the numbers here don't reflect what I see on the eGauge summary for today -- even accounting for the off by 1000 issue -- but I will fiddle with virtual registers as you suggested in the previous post first.

**Comment by jdwhite** [comment]
Actually, since I'm direct fed I'll keep the registers in my last post and show you what I'm getting for readings with the integration.  Here's a summery of today at the time I took the screenshots:

![today_summary](https://user-images.githubusercontent.com/1943283/128794491-d94a6049-861c-4d01-a613-336402cfa569.png)

Here's what the egauge "today" sensors show for solar stuff (only showing ones with last_reset attribute here):
![todays_solar_what_how](https://user-images.githubusercontent.com/1943283/128794600-331a687b-5956-4c60-afa8-cdf46daf0a7d.png)

Not even close, and again the issue with showing Wh here because they certainly aren't kWh.
The entity that does match doesn't have a last_reset attribute so I can't use it, but it's correct/accurate:

![image](https://user-images.githubusercontent.com/1943283/128794838-27a3c45e-d3f5-4cc1-8f83-52ca3761246d.png)

**Comment by neggert** [comment]
Yup, my previous post was for back-fed systems. The setup you posted looks correct for direct-feed.

I'm not sure why it would be giving a number that is off by 1000. Your numbers for grid_consumed and grid_returned look reasonable. All of the historical sensors are processed using the same code in the integration, so I don't see how some would be getting processed correctly, but not others.

I can only guess that something strange is happening on the eGauge side. Have you made changes to the register set-up since midnight? The eGauge resets some stuff when you change registers, which can lead to strange readings.

**Comment by jdwhite** [comment]
I haven't played with registers since midnight, but I did have the eGauge disconnected from mains power for a while today.

**Comment by neggert** [comment]
Hmm, that could be it. When registers get reset, you can get huge-looking numbers. The value the integration returns is the value of the counter now minus the value at start of day. If there was a reset in there, you could get a large negative number. You mentioned that your solar readings are negative, though, so that might account for the large positive number.

Side note: If you swap L1 and L2 in your solar register, it should fix your negative reading problem. You'll probably also need to adjust your "total" and "generated" virtual registers, as those have probably been configured to take into account your negative solar readings. Section 2.2 in the [eGauge config guide](https://www.egauge.net/media/support/docs/config-guide.pdf) shows what the settings should be.

**Comment by phidauex** [comment]
I'm working with the `master` version now, and have configured for backfed solar. Here are my eGauge settings (L1 and L2 got mislabeled at some point so they are largely swapped in my config):
![image](https://user-images.githubusercontent.com/49047/129093003-fdd6c17b-2d9a-4ed4-8db9-d8e4a9483dfa.png)

I get the expected positive and negative values according to eGauge conventions (positive means from the entity, negative means to the entity).

This is working for Grid In and Solar, since those are both coming in as positive values. However, the Grid Out register is giving positive values, and it looks like HASS is expecting a negative value there, therefore none of my return-to-grid is being calculated. I think this can be fixed by changing my virtual register to "- Grid-"? I will try it and see. 

![image](https://user-images.githubusercontent.com/49047/129093563-b81900b5-208d-4ad3-b92b-b08f49243066.png)

I'm also having an issue with my loads - they are correctly registering as negative values (energy TO the entity), but HASS expects positive values here. I could create eGauge virtual registers that invert the loads, but the entity list is starting to get pretty big... It would be a nice HASS addition to the Energy panel if you could select an entity, and then select "Energy FROM this entity is negative/positive" as an option, since I don't think there is one universal convention for sign on energy flow.

I'll try tweaking the Grid Out values to see if I get better results, and will report back.

**Comment by phidauex** [comment]
Ok, a few updates - things seem to be working well now with my backfeed solar configuration and measured loads. After setting up the eGauge so that the signs are correct (from the grid is positive, to the grid is negative, from the solar is positive, to loads is negative), then the trick would be to make Virtual Registers for all the devices you want to track.

Perhaps this could be a change in the code - don't create `_todays_` values for standard registers, since the signs will be wrong, but instead ask users to make their own Virtual Registers for everything they want to track in the Energy area?

My virtual registers. Note that Grid Out needs to be "minus" at the beginning to get the sign correct for HA. Also note the minus on my loads, that is just to flip the sign for HA (it wants positive numbers for consumption). The EV load register is the sum of two registers - my NEMA 14-50 outlet and my Grizzle Charger.

![image](https://user-images.githubusercontent.com/49047/129106203-e0ae0a1b-a632-492c-82f3-20a6d09f88d1.png)

Here is how the Energy panel looks now - I'm currently backfeeding solar to the grid. The totals are a little goobered since I made the Grid Out change mid-day, but I'll confirm tomorrow that I'm getting totals that match the eGauge, but I feel good about it now that the signs are right.

![image](https://user-images.githubusercontent.com/49047/129106521-ba9c7731-9259-4a98-a04e-1b28dea0ba4f.png)

Loads are looking better now that the signs are correct, though I need to wait until tomorrow to confirm because they are showing quite large values (since they haven't reset yet).

![image](https://user-images.githubusercontent.com/49047/129106752-09dc5c81-d017-479e-b0aa-d55a8a3eb798.png)

**Comment by neggert** [comment]
Glad to hear you have it working. FWIW, my value for the "return to grid" sensor is positive and that seems to be working fine for me.

I'm guessing that future release of Home Assistant will add some more flexibility to how this is set up. Right now it wants a very specific set of inputs, and it doesn't support templates. This means we have to spend a bunch of time fiddling with our hardware.

**Comment by phidauex** [comment]
In my case, Grid- generates negative values (I _think_ it would for everyone, since Grid- filters for negative values only (energy going to the entity), so `- Grid-` gets them back to positive. I think the generalized statement here is that HA wants positive values for everything, and it uses the category of the entity to determine whether it gets added or subtracted.

Thanks for all your work on this integration, even in this alpha state it seems to be working well and nice to be able to take advantage of a built-in HA feature rather than just my slapped together dashboard for power.

**Comment by jdwhite** [comment]
Now 8 minutes passed midnight CDT here (last_reset attributes reflect 2021-08-14T00:00:00-05:00) and yet all of these entities have not reset.  I believe what I though were entities reporting Wh instead of kWh (in a previous message in this thread) are just several days accumulation from not getting reset to zero nightly.

Shouldn't these entities have reset to a zero value at midnight?

![image](https://user-images.githubusercontent.com/1943283/129434746-add1aca6-6845-4848-b092-6e95e1658f6e.png)

**Comment by jdwhite** [comment]
Out of curiosity I set sensor.egauge_todays_gen to zero and it got set back to its original value shortly thereafter, so it's being calculated from other values that are rolling counters -- fine. I'm still confused why the last_reset attribute is being set at midnight then.

**Comment by neggert** [comment]
> In my case, Grid- generates negative values (I _think_ it would for everyone, since Grid- filters for negative values only (energy going to the entity), so `- Grid-` gets them back to positive. I think the generalized statement here is that HA wants positive values for everything, and it uses the category of the entity to determine whether it gets added or subtracted.
> 
> Thanks for all your work on this integration, even in this alpha state it seems to be working well and nice to be able to take advantage of a built-in HA feature rather than just my slapped together dashboard for power.

You're right, my mistake. Thanks for helping figure this out. I'm going to add all these instructions to the readme when I get time.

**Comment by neggert** [comment]
@jdwhite in general, I don't think manually setting the state using developer tools in hass does much. Manually setting it doesn't pass any calls down into the integration, so the integration just overwrites the value every time it refreshes.

Are your other sensors, e.g. `egauge_daily_use` giving sensible values?

Are the clocks and timezones accurate for both the eGauge and your Hass install? The integration isn't keeping track of anything internally, it's just querying the eGauge to compare the current value of the registers versus the value at a timestamp that's supposed to be midnight of the current day. If the date and time don't match between eGauge and Hass, it might explain the behavior you're seeing.

**Comment by jdwhite** [comment]
> @jdwhite in general, I don't think manually setting the state using developer tools in hass does much. Manually setting it doesn't pass any calls down into the integration, so the integration just overwrites the value every time it refreshes.

Agreed -- and that's exactly what my fiddling of setting the `last_reset` attribute got me so moving on.
 
> Are your other sensors, e.g. `egauge_daily_use` giving sensible values?

Yes, but they don't have the `last_reset` attribute so I can't use them as sources in the energy panel. I've worked around this by creating template sensors that are literally using the `_daily_` values with the addition of a `last_reset` attribute. Jury is still out on whether this is a viable solution.

> Are the clocks and timezones accurate for both the eGauge and your Hass install? The integration isn't keeping track of anything internally, it's just querying the eGauge to compare the current value of the registers versus the value at a timestamp that's supposed to be midnight of the current day. If the date and time don't match between eGauge and Hass, it might explain the behavior you're seeing.

Good idea, but the clocks are correct.

**Comment by jdwhite** [comment]
Here's what my `_todays_` sensors look like graphed over 7 days. As you can see, they don't reset to zero at midnight:
(ignore thew drop/recover around the start of the 14th -- that was me messing around)

![image](https://user-images.githubusercontent.com/1943283/129491423-af31a797-c1b2-4093-a41f-20ea05fa7111.png)
![image](https://user-images.githubusercontent.com/1943283/129491534-dbd9f676-d636-481e-a962-6e8ce092eccd.png)

As you can see, these values do _not_ represent current day's anything. As a quick reference, my eGauge is configured like this:

![image](https://user-images.githubusercontent.com/1943283/129492511-afe50488-4d96-4371-b0bd-ec136608ae3d.png)

As I mentioned before, since the `_daily_` values are valid I'm creating template sensors, added a `last_reset` attribute and using those in the Energy panel. 

```
  - name: "jdw_egauge_daily_grid_consumed"
    device_class: energy
    state_class: measurement
    unit_of_measurement: 'kWh'
    state: '{{ states("sensor.egauge_daily_grid_consumed") }}'
    attributes:
      last_reset: '{{ state_attr("sensor.egauge_todays_grid_consumed","last_reset") }}'
```

I am using the last_reset of the corresponding _todays_ sensor simply because it has the proper UTC offset and I didn't want to hard-code it so it would be correct when we switch on/off DST.  This setup doesn't _quite_ work because the value of my template sensor doesn't _actually_ update until the first change in the `_todays_` sensor after midnight, so technically at midnight the template sensor is still the value it was at 23:59:59:

![image](https://user-images.githubusercontent.com/1943283/129492314-95f95e7b-5107-4446-9c56-18b6c8ff4ea4.png)

and I think it caused the seriously negative solar generation at the midnight hour today:

![image](https://user-images.githubusercontent.com/1943283/129492423-b0433075-ab12-4602-b2dd-e5ceb37a2beb.png)

My thinking is that because the _last_reset_ attribute is set to 00:00:00 but isn't actually reset until the first update by the egauge integration after midnight (00:00:12 today, as shown above) that this is causing the issue. I'm going to try using an automation that fires at midnight that reset my template sensors to zero and then actually. One concern is that if it doesn't fire precisely at 00:00:00 then I could run in to the same issue, so in addition to resetting my template sensor at midnight I will set the `last_reset` value on my template sensors to the time I actually reset them instead of using the _last_reset_ on the corresponding _todays_ attribute from the egauge integration.

**Comment by neggert** [comment]
@jdwhite Let's move this discussion to a separate issue: #47. It seems like the energy dashboard integration is working for most people, so we'll need to figure out what's going on with your system.
